extension-qr-code/src/org/eclipse/birt/extension/qrcode/QRCodeBuilder.java

```diff
@@ -0,0 +1,179 @@
+/*******************************************************************************
+ * Copyright (c) 2008 Actuate Corporation.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0/.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ *
+ * Contributors:
+ *  Actuate Corporation  - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.birt.extension.qrcode;
+
+import org.eclipse.birt.report.designer.ui.dialogs.ExpressionBuilder;
+import org.eclipse.birt.report.designer.ui.dialogs.ExpressionProvider;
+import org.eclipse.birt.report.designer.ui.extensions.ReportItemBuilderUI;
+import org.eclipse.birt.report.model.api.ExtendedItemHandle;
+import org.eclipse.birt.report.model.api.extension.IReportItem;
+import org.eclipse.jface.dialogs.IDialogConstants;
+import org.eclipse.jface.dialogs.TrayDialog;
+import org.eclipse.jface.window.Window;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.SelectionAdapter;
+import org.eclipse.swt.events.SelectionEvent;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swt.widgets.Spinner;
+import org.eclipse.swt.widgets.Text;
+
+/**
+ * QRCodeBuilder
+ */
+public class QRCodeBuilder extends ReportItemBuilderUI {
+
+	@Override
+	public int open(ExtendedItemHandle handle) {
+		try {
+			IReportItem item = handle.getReportItem();
+
+			if (item instanceof QRCodeItem) {
+				QRCodeEditor editor = new QRCodeEditor(Display.getCurrent().getActiveShell(),
+						(QRCodeItem) item);
+
+				return editor.open();
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+		return Window.CANCEL;
+	}
+}
+
+/**
+ * QRCodeEditor
+ */
+class QRCodeEditor extends TrayDialog {
+
+	private QRCodeItem qrItem;
+
+	private Text txtText;
+	private Spinner spDotsWidth;
+	private Text txtEncoding;
+	private Label lbText;
+	private Label lbDotsWidth;
+	private Label lbEncoding;
+
+	protected QRCodeEditor(Shell shell, QRCodeItem qrItem) {
+		super(shell);
+
+		this.qrItem = qrItem;
+	}
+
+	@Override
+	protected void configureShell(Shell newShell) {
+		super.configureShell(newShell);
+
+		newShell.setText("QRCode Builder"); //$NON-NLS-1$
```

Comment:
Please don't mark strings that requires translation with NON-NLS

Reply:
Changed.